### PR TITLE
feat: Manage Surveys Component

### PR DESCRIPTION
**Pull request by punwai**
## Description
There is no survey component in the web app. This adds a survey component into the >System directory and routes it up.

## Related issues and discussion
#186 

## Screenshots
mifmov

screen shot 2018-12-09 at 12 36 21

## Checklist
Please make sure these boxes are checked before submitting your pull request - thanks!

- [X] If you have multiple commits please combine them into one commit by squashing them.

- [X] Read and understood the contribution guidelines at `web-app/.github/CONTRIBUTING.md`.

### CODE REVIEW

**Comment by punwai** [comment]
Fixed all the problems mentioned last commit and:
screen shot 2018-12-08 at 17 24 25

File: src/app/system/manage-surveys/manage-surveys.component.html [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    </mat-form-field>
+  </div>
+  
+  <div class="mat-elevation-z8">
+    <table mat-table [dataSource]="dataSource" matSort>
+
+      <!-- Key Column -->
+      <ng-container matColumnDef="key">
+          <th mat-header-cell *matHeaderCellDef mat-sort-header='key'> Key </th>
+          <td mat-cell *matCellDef="let survey">
```
Comment (abhaychawla) [inline src/app/system/manage-surveys/manage-surveys.component.html]: Keep all the content related to `<td>` on a single line if its just a value of an object.

File: src/app/system/manage-surveys/manage-surveys.component.html [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+      <input matInput (keyup)="applyFilter($event.target.value)">
+    </mat-form-field>
+  </div>
+  
+  <div class="mat-elevation-z8">
+    <table mat-table [dataSource]="dataSource" matSort>
+
+      <!-- Key Column -->
+      <ng-container matColumnDef="key">
+          <th mat-header-cell *matHeaderCellDef mat-sort-header='key'> Key </th>
```
Comment (abhaychawla) [inline src/app/system/manage-surveys/manage-surveys.component.html]: Don't specify a value with `mat-sort-header` anywhere. Remove the values. Also, make sure that you use only double quotes in html files unless necessary to use single quotes as well.

File: src/app/system/manage-surveys/manage-surveys.component.html [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+      </ng-container>
+  
+      <!-- Status Column -->
+      <ng-container matColumnDef="status">
+        <th mat-header-cell *matHeaderCellDef mat-sort-header='status'> Status </th>
+        <td mat-cell *matCellDef="let survey">
+            <div [className]="!isActive(survey.validFrom, survey.validTo) === true ? 'disabled' : 'enabled'">
+                <fa-icon matTooltip="{{ !isActive(survey.validFrom, survey.validTo) === true ? 'Disabled' : 'Enabled' }}" matTooltipPosition="right" icon="circle" size="lg"></fa-icon>
+            </div>
+
```
Comment (abhaychawla) [inline src/app/system/manage-surveys/manage-surveys.component.html]: Remove extra blank lines.

File: src/app/system/manage-surveys/manage-surveys.component.html [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        <td mat-cell *matCellDef="let survey"> 
+          {{ survey.countryCode }} 
+        </td>
+      </ng-container>
+  
+      <!-- Status Column -->
+      <ng-container matColumnDef="status">
+        <th mat-header-cell *matHeaderCellDef mat-sort-header='status'> Status </th>
+        <td mat-cell *matCellDef="let survey">
+            <div [className]="!isActive(survey.validFrom, survey.validTo) === true ? 'disabled' : 'enabled'">
```
Comment (abhaychawla) [inline src/app/system/manage-surveys/manage-surveys.component.html]: Update the gif/screenshot to reflect these changes.

File: src/app/system/manage-surveys/manage-surveys.component.html [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+            </div>
+    
+        </td>
+      </ng-container>
+  
+      <!-- Action Column -->
+      <ng-container matColumnDef="action">
+        <th mat-header-cell *matHeaderCellDef mat-sort-header='action'> Action </th>
+        <td mat-cell *matCellDef="let survey">
+            <a mat-raised-button [hidden]="!isActive(survey.validFrom, survey.validTo)" type="button" color="warn">Deactivate</a>
```
Comment (abhaychawla) [inline src/app/system/manage-surveys/manage-surveys.component.html]: Use button tag with an icon to ensure consistency with other parts of the application for example:
```html
<button mat-button color="accent">
    <fa-icon icon="lock-open"></fa-icon>&nbsp;&nbsp;Activate
</button>
<button mat-button color="warn">
    <fa-icon icon="lock"></fa-icon>&nbsp;&nbsp;Deactivate
</button>
```

Also, don't use `hidden` use `*ngIf` instead.

File: src/app/system/manage-surveys/manage-surveys.component.html [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+            <a mat-raised-button [hidden]="isActive(survey.validFrom, survey.validTo)" type="button" color="accent">Activate</a>
+        </td> 
+      </ng-container>
+  
+      <tr mat-header-row *matHeaderRowDef="displayedColumns"></tr>
+      <tr mat-row *matRowDef="let survey; columns: displayedColumns;" class="select-row"></tr>
+    </table>
+
+    <mat-paginator [pageSizeOptions]="[10, 25, 50, 100]"  showFirstLastButtons></mat-paginator>
```
Comment (abhaychawla) [inline src/app/system/manage-surveys/manage-surveys.component.html]:
```suggestion
    <mat-paginator [pageSizeOptions]="[10, 25, 50, 100]" showFirstLastButtons></mat-paginator>
```

File: src/app/system/manage-surveys/manage-surveys.component.html [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+<div class="container">
+
+  <div fxLayout="row">
+    <mat-form-field fxFlex>
+      <mat-label>Filter</mat-label>
+      <input matInput (keyup)="applyFilter($event.target.value)">
+    </mat-form-field>
+  </div>
+  
+  <div class="mat-elevation-z8">
```
Comment (abhaychawla) [inline src/app/system/manage-surveys/manage-surveys.component.html]: Insert a blank line after this.

File: src/app/system/manage-surveys/manage-surveys.component.html [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+      <ng-container matColumnDef="action">
+        <th mat-header-cell *matHeaderCellDef mat-sort-header='action'> Action </th>
+        <td mat-cell *matCellDef="let survey">
+            <a mat-raised-button [hidden]="!isActive(survey.validFrom, survey.validTo)" type="button" color="warn">Deactivate</a>
+            <a mat-raised-button [hidden]="isActive(survey.validFrom, survey.validTo)" type="button" color="accent">Activate</a>
+        </td> 
+      </ng-container>
+  
+      <tr mat-header-row *matHeaderRowDef="displayedColumns"></tr>
+      <tr mat-row *matRowDef="let survey; columns: displayedColumns;" class="select-row"></tr>
```
Comment (abhaychawla) [inline src/app/system/manage-surveys/manage-surveys.component.html]: Insert a blank line after this.

File: src/app/system/manage-surveys/manage-surveys.component.html [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+    </table>
+
+    <mat-paginator [pageSizeOptions]="[10, 25, 50, 100]"  showFirstLastButtons></mat-paginator>
+
+  </div>
```
Comment (abhaychawla) [inline src/app/system/manage-surveys/manage-surveys.component.html]: Insert a blank line after this.

File: src/app/system/manage-surveys/manage-surveys.component.html (hunk 8 above)
Comment (abhaychawla) [inline src/app/system/manage-surveys/manage-surveys.component.html]: select-row class doesn't exist in scss file.

File: src/app/system/manage-surveys/manage-surveys.component.scss [hunk 10]
```diff
@@ -0,0 +1,19 @@
+table {
+  width: 100%;
+}
+
+[hidden] {
```
Comment (abhaychawla) [inline src/app/system/manage-surveys/manage-surveys.component.scss]: Remove this class.

File: src/app/system/manage-surveys/manage-surveys.component.scss [hunk 11]
```diff
@@ -0,0 +1,19 @@
...
+
+.enabled {
+  color: #32cd32;
+}
+
+.disabled {
+  color: #f44366;
+}
+
+td, th {
```
Comment (abhaychawla) [inline src/app/system/manage-surveys/manage-surveys.component.scss]: Remove this scss.

File: src/app/system/manage-surveys/manage-surveys.component.ts [hunk 12]
```diff
@@ -0,0 +1,68 @@
+/** Angular Imports */
+import { Component, OnInit, ViewChild } from '@angular/core';
+import { MatTableDataSource, MatPaginator, MatSort } from '@angular/material';
+import { ActivatedRoute } from '@angular/router';
+
```
Comment (abhaychawla) [inline src/app/system/manage-surveys/manage-surveys.component.ts]: Mention all comments. Take reference from `accounting` section.

File: src/app/system/manage-surveys/manage-surveys.component.ts [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+   */
+  constructor(private route: ActivatedRoute) {
+    this.route.data.subscribe(( data: { surveys: any }) => {
+      this.surveysData = data.surveys;
+    });
+  }
+
+
+  /**
+   * Returns whether an element is active based on its active dates.
```
Comment (abhaychawla) [inline src/app/system/manage-surveys/manage-surveys.component.ts]: Change element to `survey` and `based on its duration`.

File: src/app/system/manage-surveys/manage-surveys.component.ts [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+   * @param {number} validFrom Date valid from
+   * @param {number} validTo Date valid to
+   */
+  isActive = function( validFrom: number, validTo: number) {
```
Comment (abhaychawla) [inline src/app/system/manage-surveys/manage-surveys.component.ts]:
```suggestion
  isActive(validFrom: number, validTo: number) {
```

Is function keyword necessary?

File: src/app/system/manage-surveys/manage-surveys.component.ts [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+    const curdate = new Date().getTime();
+    return (curdate > validFrom && curdate < validTo);
+  };
+
+  ngOnInit() {
+    this.getSurveys();
```
Comment (abhaychawla) [inline src/app/system/manage-surveys/manage-surveys.component.ts]: Change to setSurveys.

File: src/app/system/manage-surveys/manage-surveys.component.ts [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+      switch (property) {
+        case 'status': return this.isActive(item.validFrom, item.validTo);
+        case 'action': return this.isActive(item.validFrom, item.validTo);
+        default: return item[property];
+      }
+    };
+  }
+
+  /**
+   * Initializes the data source, paginator and sorter for codes table.
```
Comment (abhaychawla) [inline src/app/system/manage-surveys/manage-surveys.component.ts]: Incorrect comment.

**Comment by punwai** [comment]
demodemo
This is the new demo

**Comment by abhaychawla** [comment]
@punwai If the status is `active` then the action should be `Deactivate` and vice versa. They have been swapped.

**Comment by punwai** [comment]
@abhaychawla 
screen shot 2018-12-09 at 12 14 57
Fixed and pushed

**Comment by abhaychawla** [comment]
> Also aligned Action title with buttons
> screen shot 2018-12-09 at 12 24 01

No, remove this alignment. It's not needed. Also resolve all the conversation above (click on resolve button) so I can take a look. Put the final screenshot in the main description at the top.

**Comment by punwai** [comment]
@abhaychawla Removed the alignment + Added a screenshot at the top

File: src/app/system/manage-surveys/manage-surveys.component.html [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+      </ng-container>
+
+      <ng-container matColumnDef="action">
+        <th mat-header-cell *matHeaderCellDef mat-sort-header> Action </th>
+        <td mat-cell *matCellDef="let survey">
+          <button mat-button *ngIf="!isActive(survey.validFrom, survey.validTo)" color="accent">
+            <fa-icon icon="lock-open"></fa-icon>&nbsp;&nbsp;Activate
+          </button>
+          <button mat-button *ngIf="isActive(survey.validFrom, survey.validTo)" color="warn">
+              <fa-icon icon="lock"></fa-icon>&nbsp;&nbsp;Deactivate
```
Comment (abhaychawla) [inline src/app/system/manage-surveys/manage-surveys.component.html]: Ensure its indented with the same amount of space as above.

File: src/app/system/manage-surveys/manage-surveys.component.scss [hunk 18]
```diff
@@ -0,0 +1,20 @@
+table {
```
Comment (abhaychawla) [inline src/app/system/manage-surveys/manage-surveys.component.scss]: Redundant as its already specified below, remove this.

File: src/app/system/manage-surveys/manage-surveys.component.scss [hunk 19]
```diff
@@ -0,0 +1,20 @@
...
+}
+
+// TODO: CONFIGURE THESE CLASSES GLOBALLY
+.enabled {
+  color: #32cd32;
+}
+
+.disabled {
+  color: #f44366;
+}
```
Comment (abhaychawla) [inline src/app/system/manage-surveys/manage-surveys.component.scss]: Ensure file ends with a single blank line.

File: src/app/system/manage-surveys/manage-surveys.component.ts [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+  /* Data source for manage surveys data table */
+  dataSource: MatTableDataSource<any>;
+
+  /** Paginator for manage surveys table. */
+  @ViewChild(MatPaginator) paginator: MatPaginator;
+  /** Sorter for manage surveys table. */
+  @ViewChild(MatSort) sort: MatSort;
+
+  /**
+   * Retrieves the codes data from `resolve`.
```
Comment (abhaychawla) [inline src/app/system/manage-surveys/manage-surveys.component.ts]: Incorrect comment.

Comment (punwai) [reply src/app/system/manage-surveys/manage-surveys.component.ts]: pushed

File: src/app/system/manage-surveys/manage-surveys.component.scss (hunk 19 above)
Comment (punwai) [reply src/app/system/manage-surveys/manage-surveys.component.scss]: pushed

**Comment by punwai** [comment]
@abhaychawla Fixed problems and pushed.

**Pull request merged**